docs/exceptions.md

```diff
@@ -0,0 +1,266 @@
+you're not aware of. Always make this trade-off conciously and responsibly.
+
+(On the other hand, if static guarantees require GADTs or type families, perhaps
+it's better to not overcomplicate code. Use your judgement.)
+
+DISCUSSION: Should we create a synonym `bug = impureThrow` in Universum? This
+would make the intention more clear.
+
+### Pure code, regular errors
+
+Do *not*:
+
+* use `error` or `impureThrow`
+* use `MonadFail`
+* return `Either Text`
+
+Do:
+
+* return `Either ErrorADT`, `Maybe`
+* wrap the underlying (pure!) monad in `ExceptT` or `CatchT`
+* use `MonadError` or `MonadThrow`
+
+Consider parsing: it is pure, but we cannot make assumptions about the input. In
+this case we might want to use `ExceptT ParseError`. Or consider a lookup in a
+`Map`, where we don't know whether the key is present -- in this case we'd like
+to return `Maybe v`. In 100% pure code, use one of these ways to handle errors:
+
+* `Maybe`, `Either e`
+* `MaybeT`, `ExceptT e`
+* `CatchT`
+
+Avoid using `Text` with the error message in place of `e` -- create a proper
+ADT. In case creating a proper ADT feels too cumbersome, use `CatchT`, which
+is equivalent to `ExceptT SomeException`.
+
+Be careful not to use `MaybeT`, `ExceptT`, and `CatchT` in potentially impure
+code. When in doubt whether the code is potentially impure, use `MonadThrow`.  
+(The reason we don't want `ExceptT` and co. in potentially impure code is that
+they add additional exception mechanisms to the one that `IO` has, and
+`catch`/`bracket` don't account for this).
+
+### Impure code, regular errors
+
+Do *not*:
+
+* use `error` or `impureThrow`
+* use `ExceptT`, `MaybeT`, or `CatchT`
+* use `Maybe` or `Either`
+* use `MonadError`
+* use `throwIO`
+
+Do:
+
+* create a custom exception type
+* use `throwM` (`MonadThrow`)
+
+We disallow the use of `throwIO` only because it is redundant in the presence of
+`throwM` and requires a stronger constraint (`MonadIO` rather than
+`MonadThrow`).
+
+Derive prisms for exception types with multiple constructors, so it's convenient
+to use them with `catchJust`.
+
+### Impure code, programmer mistakes
+
+Use the same techniques as in pure code -- `error` or `impureThrow`. There are
+two reasons for this:
+
+* we would rather catch the error sooner than later, and `impureThrow` explodes
+  when forced to WHNF, while `throwIO` or `throwM` explode when executed
+
+* when in an abstract (but potentially impure) monad, using `throwM` might
+  add an additional constraint
+
+## Packages and modules
+
+Do not import `Control.Exception` or `Control.Monad.Catch`! We use the
+`safe-exceptions` to deal with asynchronous exceptions gracefully, so import
+`Control.Exception.Safe`.
+
+## Resource handling
+
+Use `bracket` or to guarantee the release of resources. In case of
+concurrent code, avoid `forkIO` or `forkProcess` in favor of the `async`
+package, as it rethrows exceptions from the child threads.
```

Comment:
Yes and prefer `withAsync` / `concurrently` / `race` wherever possible over `async` itself, because that one will *not* automatically kill the child if the parent dies. `withAsync` basically does a `bracket` so don't use it in a tail-recursive loop like `loop = waitForData >>= \it -> withAsync (process it) >> loop`, and be aware of `link` in case you don't plan to `wait` for the `Async`.

Reply:
Yes, that's the way to use `async` that I had in mind. I added a clarification.